src/src/dynpower.cpp

```diff
@@ -0,0 +1,202 @@
+#if defined(TARGET_TX)
+
+#include <dynpower.h>
+#include <common.h>
+
+// LQ-based boost defines
+#define DYNPOWER_LQ_BOOST_THRESH_DIFF 20  // If LQ is dropped suddenly for this amount (relative), immediately boost to the max power configured.
+#define DYNPOWER_LQ_BOOST_THRESH_MIN  50  // If LQ is below this value (absolute), immediately boost to the max power configured.
+#define DYNPOWER_LQ_MOVING_AVG_K      8   // Number of previous values for calculating moving average. Best with power of 2.
+#define DYNPOWER_LQ_THRESH_UP         85  // Below this LQ, the RSSI/SNR code will increase the power if RSSI/SNR did nothing
+
+// RSSI-based increment defines
+#define DYNPOWER_RSSI_CNT 5               // Number of RSSI readings to average (straight average) to make an RSSI-based adjustment
+#define DYNPOWER_RSSI_THRESH_UP 15        // RSSI < (Sensitivity+Up) -> raise power
+#define DYNPOWER_RSSI_THRESH_DN 21        // RSSI > (Sensitivity+Dn) >- lower power
+
+// SNR-based increment defines
+#define DYNPOWER_LQ_THRESH_DN 95          // Min LQ for lowering power using SNR-based power lowering
+
+template<uint8_t K, uint8_t SHIFT>
+class MovingAvg
+{
+public:
+  void init(uint32_t v) { _shiftedVal = v << SHIFT; };
+  void add(uint32_t v) {  _shiftedVal = ((K - 1) * _shiftedVal + (v << SHIFT)) / K; };
+  uint32_t getValue() const { return _shiftedVal >> SHIFT; };
+
+  void operator=(const uint32_t &v) { init(v); };
+  operator uint32_t () const { return getValue(); };
+private:
+  uint32_t _shiftedVal;
+};
+
+static MovingAvg<DYNPOWER_LQ_MOVING_AVG_K, 16> dynpower_mavg_lq;
+static MeanAccumulator<int32_t, int8_t, -128> dynpower_mean_rssi;
+static int8_t dynpower_updated;
+static uint32_t dynpower_last_linkstats_millis;
+
+static void DynamicPower_SetToConfigPower()
+{
+    POWERMGNT::setPower((PowerLevels_e)config.GetPower());
+}
+
+void DynamicPower_Init()
+{
+    dynpower_mavg_lq = 100;
+    dynpower_updated = DYNPOWER_UPDATE_NOUPDATE;
+}
+
+void ICACHE_RAM_ATTR DynamicPower_TelemetryUpdate(int8_t snrScaled)
+{
+    dynpower_updated = snrScaled;
+    // static unsigned lastUp;
+    // if (snrScaled != DYNPOWER_UPDATE_MISSED)
+    // {
+    //   unsigned now = millis();
+    //   DBGLN(" %u P=%u SNR=%d ", now - lastUp, POWERMGNT::currPower(), snrScaled);
+    //   lastUp = now;
+    // }
```

Comment:
Did you want to wrap this in some DEBUG_ flag? Or just remove it?

Reply:
I'll just remove it, I didn't want to have to come up with a new DEBUG_ name and it is easy enough to just throw in DBG if needed. This already served its purpose.